metrics-exporter-log/src/lib.rs

```diff
@@ -7,35 +7,34 @@
 //! # Run Modes
 //! - `run` can be used to block the current thread, taking snapshots and exporting them on an
 //! interval
-//! - `turn` can be used to take a single snapshot and log it
 //! - `into_future` will return a [`Future`] that when driven will take a snapshot on the 
 //! configured interval and log it
 #[macro_use]
 extern crate log;
 
 use std::thread;
 use std::time::Duration;
-use metrics::Controller;
-use metrics_core::MetricsRecorder;
+use metrics_core::{Recorder, Snapshot, SnapshotProvider, AsyncSnapshotProvider};
 use log::Level;
 use futures::prelude::*;
 use tokio_timer::Interval;
 
 /// Exports metrics by converting them to a textual representation and logging them.
-pub struct LogExporter<R> {
-    controller: Controller,
+pub struct LogExporter<C, R> {
+    controller: C,
     recorder: R,
     level: Level,
 }
 
-impl<R> LogExporter<R>
+impl<C, R> LogExporter<C, R>
 where
-    R: MetricsRecorder + Clone + Into<String>
+    C: SnapshotProvider + AsyncSnapshotProvider,
```

Comment:
can we move these bounds to `turn` and `into_future`?

Reply:
Is there any practical reason to do this?  It seems like it's 50/50 on whether types constraint at the type level or at the function level; I notice Tokio/Tower prefer the some of "late binding" constraint approach.

Reply:
well, in this case, someone could pass a non-async one and still use this.

Reply:
Good point, yeah, letting people write a non-async provider if they don't need async and vise versa.  Will do this.